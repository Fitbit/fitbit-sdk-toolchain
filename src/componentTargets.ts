interface ComponentTarget {
  inputs: string[];
  outputDir?: string;
  notFoundIsFatal: boolean;
  translationsGlob: string;
}

export enum ComponentType {
  DEVICE = 'device',
  COMPANION = 'companion',
  SETTINGS = 'settings',
}

const componentTarget: { [component in ComponentType]: ComponentTarget } = {
  [ComponentType.DEVICE]: {
    inputs: ['app/index.ts', 'app/index.js'],
    outputDir: 'app',
    notFoundIsFatal: true,
    translationsGlob: 'app/i18n/*.po',
  },
  [ComponentType.COMPANION]: {
    inputs: ['companion/index.ts', 'companion/index.js'],
    notFoundIsFatal: false,
    translationsGlob: 'companion/i18n/*.po',
  },
  [ComponentType.SETTINGS]: {
    inputs: ['tsx', 'ts', 'jsx', 'js'].map((ext) => `settings/index.${ext}`),
    notFoundIsFatal: false,
    translationsGlob: 'settings/i18n/*.po',
  },
};

export default componentTarget;
