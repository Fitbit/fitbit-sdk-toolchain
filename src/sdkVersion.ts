import semver from 'semver';

import packageVersionConst from './packageVersion.const';

export default function sdkVersion(toolchainVersion = packageVersionConst) {
  const version = semver.parse(toolchainVersion);
  if (version === null) {
    throw new Error(`Invalid SDK package version: ${toolchainVersion}`);
  }
  return version;
}

export function apiVersions(
  { enableProposedAPI }: { enableProposedAPI?: boolean } = {},
  toolchainVersion = packageVersionConst,
) {
  if (enableProposedAPI) return { deviceApi: '*', companionApi: '*' };

  const { major, minor } = sdkVersion(toolchainVersion);
  if (major === 1 && minor === 0) {
    return { deviceApi: '1.0.0', companionApi: '1.0.0' };
  }
  if (major === 2 && minor === 0) {
    return { deviceApi: '3.0.0', companionApi: '2.0.0' };
  }
  if (major === 3 && minor === 0) {
    return { deviceApi: '4.0.0', companionApi: '2.1.0' };
  }
  throw new Error(
    `No known API versions for SDK package version ${major}.${minor}`,
  );
}
