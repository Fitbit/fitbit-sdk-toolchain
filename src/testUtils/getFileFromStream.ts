import stream from 'stream';

import Vinyl from 'vinyl';

export default function getFileFromStream(stream: stream.Readable, filename?: string): Promise<Vinyl> {
    return new Promise<Vinyl>((resolve, reject) => {
        stream.on('data', (file: Vinyl) => {
            if (file.contents && (!filename || file.basename === filename)) {
                resolve(file);
            }
        });
        stream.on('finish', () => reject(new Error('No file in stream matched')));
        stream.on('error', reject);
    });
}
