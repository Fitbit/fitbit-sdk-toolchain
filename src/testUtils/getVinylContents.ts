import concatStream from 'concat-stream';
import Vinyl from 'vinyl';

export default function getVinylContents(file: Vinyl) {
  return new Promise<string>((resolve, reject) => {
    if (file.isBuffer()) resolve(file.contents.toString());
    else if (file.isStream()) {
      file.contents.pipe(concatStream({ encoding: 'string' }, (str) => {
        // Typing for concatStream is broken, thinks it always returns buffers
        resolve(str as any);
      }));
    }
    else reject(new Error(`Unknown Vinyl file type for path: ${file.path}`));
  });
}
