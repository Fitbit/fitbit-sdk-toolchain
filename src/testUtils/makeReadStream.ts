import { Readable } from 'stream';

export default function makeReadStream() {
  const stream = new Readable({ objectMode: true });
  stream._read = () => {};
  return stream;
}
