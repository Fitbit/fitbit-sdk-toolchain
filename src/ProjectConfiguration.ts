import humanizeList from 'humanize-list';
import { isHexColor, isUUID } from 'validator';
import lodash from 'lodash';
import semver from 'semver';

import buildTargets from './buildTargets';
import DiagnosticList from './DiagnosticList';
import { normalizeLanguageTag } from './languageTag';
import sdkVersion from './sdkVersion';

const knownBuildTargets = Object.keys(buildTargets);

export enum AppType {
  APP = 'app',
  CLOCKFACE = 'clockface',
}

export const VALID_APP_TYPES = Object.values(AppType);

export type LocalesConfig = { [locale: string]: { name: string } };

export interface BaseProjectConfiguration {
  appType: AppType;
  appDisplayName: string;
  appUUID: string;
  requestedPermissions: string[];
  i18n: LocalesConfig;
  defaultLocale: string;
  buildTargets: string[];
  // We don't want to accidentally serialize `enableProposedAPI: false`
  // out to users' package.json files.
  enableProposedAPI?: true;
}

export interface AppProjectConfiguration extends BaseProjectConfiguration {
  appType: AppType.APP;
  wipeColor: string;
  iconFile: string;
}

export interface ClockProjectConfiguration extends BaseProjectConfiguration {
  appType: AppType.CLOCKFACE;
}

type ProjectConfiguration = AppProjectConfiguration | ClockProjectConfiguration;
export default ProjectConfiguration;

export const MAX_DISPLAY_NAME_LENGTH = 30;

export const LOCALES = Object.freeze({
  en: 'English',
  de: 'German',
  es: 'Spanish',
  fr: 'French',
  it: 'Italian',
  ja: 'Japanese',
  ko: 'Korean',
  nl: 'Dutch',
  sv: 'Swedish',
  'zh-cn': 'Chinese (S)',
  'zh-tw': 'Chinese (T)',
});

const permissionTypes = [
  {
    key: 'access_activity',
    name: 'Activity',
    // tslint:disable-next-line:max-line-length
    description:
      'Read user activities for today (distance, calories, steps, elevation and active minutes), and daily goals.',
  },
  {
    key: 'access_user_profile',
    name: 'User Profile',
    // tslint:disable-next-line:max-line-length
    description:
      'Read non-identifiable personal information (gender, age, height, weight, resting HR, basal metabolic rate, stride, HR zones).',
  },
  {
    key: 'access_heart_rate',
    name: 'Heart Rate',
    description: 'Application may read the heart-rate sensor in real-time.',
  },
  {
    key: 'access_location',
    name: 'Location',
    description: 'Application and companion may use GPS.',
  },
  {
    key: 'access_internet',
    name: 'Internet',
    description:
      'Companion may communicate with the Internet using your phone data connection.',
  },
  {
    key: 'run_background',
    name: 'Run in background',
    description:
      'Companion may run even when the application is not actively in use.',
  },
  {
    key: 'access_exercise',
    name: 'Exercise Tracking',
    description: 'Application may track an exercise.',
    sdkVersion: '>=3.0.0',
  },
];

const restrictedPermissionTypes = [
  {
    key: 'fitbit_token',
    name: '[Restricted] Fitbit Token',
    description: 'Access Fitbit API token.',
  },
  {
    key: 'external_app_communication',
    name: '[Restricted] External Application Communication',
    description:
      'Allows communication between external mobile applications and companion.',
  },
  {
    key: 'access_secure_exchange',
    name: '[Restricted] Secure Exchange',
    description: 'Allows securing any data and verifying that data was secured',
  },
];

function getAllPermissionTypes() {
  return [
    ...restrictedPermissionTypes,
    ...permissionTypes.filter(
      (permission) =>
        !permission.sdkVersion ||
        semver.satisfies(sdkVersion(), permission.sdkVersion),
    ),
  ];
}

function constrainedSetDiagnostics({
  actualValues,
  knownValues,
  valueTypeNoun,
  notFoundIsFatal = false,
}: {
  actualValues: ReadonlyArray<any>;
  knownValues: ReadonlyArray<any>;
  valueTypeNoun: string;
  notFoundIsFatal?: boolean;
}) {
  const unknownValues = lodash.without(actualValues, ...knownValues);
  const diagnostics = new DiagnosticList();

  if (unknownValues.length > 0) {
    const unknownValueStrings = unknownValues.filter(lodash.isString);
    const unknownValueOther = lodash
      .without(unknownValues, ...unknownValueStrings)
      .map(String);
    if (unknownValueStrings.length) {
      const errStr = `One or more ${valueTypeNoun} was invalid: ${unknownValueStrings.join(
        ', ',
      )}`;
      if (notFoundIsFatal) diagnostics.pushFatalError(errStr);
      else diagnostics.pushWarning(errStr);
    }
    if (unknownValueOther.length) {
      diagnostics.pushFatalError(
        `One or more ${valueTypeNoun} was not a string: ${unknownValueOther.join(
          ', ',
        )}`,
      );
    }
  }

  const duplicatedValues = lodash
    .uniq(actualValues)
    .filter(
      (value) =>
        actualValues.indexOf(value) !== actualValues.lastIndexOf(value),
    );
  if (duplicatedValues.length > 0) {
    diagnostics.pushWarning(
      `One or more ${valueTypeNoun} was specified multiple times: ${duplicatedValues.join(
        ', ',
      )}`,
    );
  }

  return diagnostics;
}

/**
 * Normalize the project configuration from a parsed package.json.
 *
 * @param config Configuration object
 * @param defaults Override the default values to use
 */
export function normalizeProjectConfig(
  config: any,
  defaults?: Partial<AppProjectConfiguration & ClockProjectConfiguration>,
): ProjectConfiguration {
  if (!lodash.isPlainObject(config)) {
    throw new TypeError('Project configuration root must be an object');
  }

  const mergedConfig: ProjectConfiguration = {
    // The *cough* default defaults
    appUUID: '',
    appType: AppType.APP,
    appDisplayName: '',
    iconFile: 'resources/icon.png',
    wipeColor: '',
    requestedPermissions: [],
    buildTargets: [],
    i18n: {},
    defaultLocale: 'en-US',

    // Override defaults
    ...defaults,

    // The config object proper
    ...(config.fitbit as {}),
  };

  const normalizedDefaultLocale = normalizeLanguageTag(
    mergedConfig.defaultLocale,
  );
  if (normalizedDefaultLocale !== null) {
    mergedConfig.defaultLocale = normalizedDefaultLocale;
  }

  const { requestedPermissions } = mergedConfig;
  if (!Array.isArray(requestedPermissions)) {
    // tslint:disable-next-line:max-line-length
    throw new TypeError(
      `fitbit.requestedPermissions must be an array, not ${typeof requestedPermissions}`,
    );
  }

  return mergedConfig;
}

export function validateAppType(config: ProjectConfiguration) {
  const diagnostics = new DiagnosticList();

  if (VALID_APP_TYPES.indexOf(config.appType) === -1) {
    const appTypeNames = humanizeList(VALID_APP_TYPES, { conjunction: 'or' });
    diagnostics.pushFatalError(
      `App type '${config.appType}' is invalid, expected ${appTypeNames}`,
    );
  }
  return diagnostics;
}

export function validateDisplayName(name: string) {
  if (name.length === 0) {
    return 'Display name must not be blank';
  }
  if (name.length > MAX_DISPLAY_NAME_LENGTH) {
    return `Display name must not exceed ${MAX_DISPLAY_NAME_LENGTH} characters`;
  }
  return true;
}

export function validateProjectDisplayName(config: ProjectConfiguration) {
  const diagnostics = new DiagnosticList();
  const result = validateDisplayName(config.appDisplayName);
  if (result !== true) {
    diagnostics.pushFatalError(result);
  }
  return diagnostics;
}

export function validateWipeColor(config: ProjectConfiguration) {
  const diagnostics = new DiagnosticList();
  if (config.appType !== AppType.CLOCKFACE && !isHexColor(config.wipeColor)) {
    diagnostics.pushFatalError('Wipe color must be a valid hex color');
  }
  return diagnostics;
}

export function validateRequestedPermissions({
  requestedPermissions,
}: ProjectConfiguration) {
  return constrainedSetDiagnostics({
    actualValues: requestedPermissions,
    knownValues: getAllPermissionTypes().map((permission) => permission.key),
    valueTypeNoun: 'requested permissions',
    notFoundIsFatal: false,
  });
}

export function validateBuildTarget({ buildTargets }: ProjectConfiguration) {
  const diagnostics = constrainedSetDiagnostics({
    actualValues: buildTargets,
    knownValues: knownBuildTargets,
    valueTypeNoun: 'build targets',
    notFoundIsFatal: true,
  });

  if (buildTargets.length === 0) {
    diagnostics.pushFatalError('At least one build target must be enabled');
  }

  return diagnostics;
}

export function validateLocaleDisplayName(
  { i18n }: ProjectConfiguration,
  localeKey: keyof typeof LOCALES,
) {
  const diagnostics = new DiagnosticList();
  const locale = i18n[localeKey];

  if (!locale) return diagnostics;

  if (!locale.name || locale.name.length === 0) {
    // tslint:disable-next-line:max-line-length
    diagnostics.pushFatalError(
      `Localized display name for ${LOCALES[localeKey]} must not be blank`,
    );
  }
  if (locale.name.length > MAX_DISPLAY_NAME_LENGTH) {
    // tslint:disable-next-line:max-line-length
    diagnostics.pushFatalError(
      `Localized display name for ${
        LOCALES[localeKey]
      } must not exceed ${MAX_DISPLAY_NAME_LENGTH} characters`,
    );
  }

  return diagnostics;
}

export function validateLocaleDisplayNames(config: ProjectConfiguration) {
  const diagnostics = new DiagnosticList();
  for (const localeKey of Object.keys(LOCALES)) {
    diagnostics.extend(
      validateLocaleDisplayName(config, localeKey as keyof typeof LOCALES),
    );
  }
  return diagnostics;
}

export function validateSupportedLocales({ i18n }: ProjectConfiguration) {
  const diagnostics = new DiagnosticList();

  const unknownLocales = lodash.without(
    Object.keys(i18n),
    ...Object.keys(LOCALES),
  );
  if (unknownLocales.length > 0) {
    diagnostics.pushWarning(`Invalid locales: ${unknownLocales.join(', ')}`);
  }

  return diagnostics;
}

export function validateAppUUID({ appUUID }: ProjectConfiguration) {
  const diagnostics = new DiagnosticList();
  if (!isUUID(String(appUUID))) {
    diagnostics.pushFatalError(
      'appUUID must be a valid UUID, run "npx fitbit-build generate-appid" to fix',
    );
  }

  return diagnostics;
}

export function validateDefaultLocale(config: ProjectConfiguration) {
  const diagnostics = new DiagnosticList();
  if (normalizeLanguageTag(config.defaultLocale) === null) {
    diagnostics.pushFatalError(
      `Default locale is an invalid language tag: ${config.defaultLocale}`,
    );
  }
  return diagnostics;
}

export function validate(config: ProjectConfiguration) {
  const diagnostics = new DiagnosticList();
  [
    validateAppUUID,
    validateProjectDisplayName,
    validateAppType,
    validateWipeColor,
    validateRequestedPermissions,
    validateBuildTarget,
    validateSupportedLocales,
    validateLocaleDisplayNames,
    validateDefaultLocale,
  ].forEach((validator) => diagnostics.extend(validator(config)));
  return diagnostics;
}
