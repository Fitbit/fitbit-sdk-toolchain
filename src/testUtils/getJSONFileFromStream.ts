import { Readable } from 'stream';

import getFileFromStream from './getFileFromStream';
import getVinylContents from './getVinylContents';

export default function getJSONFileFromStream(
  stream: Readable,
  filename?: string,
) {
  return getFileFromStream(stream, filename)
    .then(getVinylContents)
    .then(JSON.parse);
}
