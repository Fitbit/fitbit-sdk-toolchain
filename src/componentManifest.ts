import { Transform } from 'stream';

import gulpFile from 'gulp-file';
import Vinyl from 'vinyl';

import componentTargets from './componentTargets';
import { normalizeToPOSIX } from './pathUtils';
import ProjectConfiguration, { AppType } from './ProjectConfiguration';
import * as resources from './resources';
import { apiVersions } from './sdkVersion';

const manifestPath = 'manifest.json';

interface ComponentManifest {
  apiVersion: string;
  buildId: string;
  bundleDate: string;
  name: string;
  requestedPermissions?: string[];
  uuid: string;
}

interface DeviceManifest extends ComponentManifest {
  appManifestVersion: 1;
  // Enum values are hardcoded instead of reusing AppType since this
  // interface describes the manifest as the device expects it,
  // independent of data structures elsewhere. We want the build to
  // fail if the AppType enum changes such that it becomes incompatible
  // with what the device expects.
  appType: 'app' | 'clockface';
  i18n: {
    [locale: string]: {
      name?: string;
      resources?: string;
    };
  };
  iconFile?: string;
  main: string;
  svgMain: string;
  svgWidgets: string;
  wipeColor?: string;
}

interface CompanionManifest extends ComponentManifest {
  manifestVersion: 2;
  companion: {
    main: string;
  };
  settings?: {
    main: string;
  };
}

function makeCommonManifest({
  projectConfig,
  buildId,
  apiVersion,
}: {
  projectConfig: ProjectConfiguration;
  buildId: string;
  apiVersion: string;
}): ComponentManifest {
  return {
    apiVersion,
    buildId,
    bundleDate: new Date().toISOString(),
    uuid: projectConfig.appUUID,
    name: projectConfig.appDisplayName,
    requestedPermissions: projectConfig.requestedPermissions,
  };
}

export function makeDeviceManifest({
  projectConfig,
  buildId,
}: {
  projectConfig: ProjectConfiguration;
  buildId: string;
}) {
  const manifest: DeviceManifest = {
    appManifestVersion: 1,
    main: componentTargets.device.output,
    svgMain: resources.svgMain,
    svgWidgets: resources.svgWidgets,
    appType: projectConfig.appType,
    i18n: projectConfig.i18n,
    ...makeCommonManifest({
      projectConfig,
      buildId,
      apiVersion: apiVersions(projectConfig).deviceApi,
    }),
  };

  if (projectConfig.appType !== AppType.CLOCKFACE) {
    manifest.iconFile = projectConfig.iconFile;
    manifest.wipeColor = projectConfig.wipeColor;
  }

  return new Transform({
    objectMode: true,

    transform(file: Vinyl, _, next) {
      const lang: string | undefined = file.translationLanguage;

      if (lang) {
        if (manifest.i18n[lang] === undefined) manifest.i18n[lang] = {};
        manifest.i18n[lang].resources = normalizeToPOSIX(file.relative);
      }

      next(undefined, file);
    },

    flush(done) {
      // Ensure the default locale is the first listed in the manifest
      const {
        [projectConfig.defaultLocale]: defaultLocale,
        ...otherLocales
      } = manifest.i18n;
      manifest.i18n = {
        [projectConfig.defaultLocale]: defaultLocale,
        ...otherLocales,
      };

      done(
        undefined,
        new Vinyl({
          cwd: '',
          base: undefined,
          path: manifestPath,
          contents: Buffer.from(JSON.stringify(manifest)),
        }),
      );
    },
  });
}

export function makeCompanionManifest({
  projectConfig,
  hasSettings,
  buildId,
}: {
  projectConfig: ProjectConfiguration;
  hasSettings: boolean;
  buildId: string;
}) {
  const manifest: CompanionManifest = {
    manifestVersion: 2,
    companion: { main: componentTargets.companion.output },
    ...makeCommonManifest({
      projectConfig,
      buildId,
      apiVersion: apiVersions(projectConfig).companionApi,
    }),
  };

  if (hasSettings) {
    manifest.settings = { main: componentTargets.settings.output };
  }
  return gulpFile(manifestPath, JSON.stringify(manifest));
}
