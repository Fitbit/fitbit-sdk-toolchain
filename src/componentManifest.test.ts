import { Duplex, Readable } from 'stream';

import { advanceTo } from 'jest-date-mock';
import path from 'path';
import Vinyl from 'vinyl';

import { makeDeviceManifest, makeCompanionManifest } from './componentManifest';
import { ComponentType } from './componentTargets';
import ProjectConfiguration, {
  AppProjectConfiguration,
  AppType,
  ClockProjectConfiguration,
} from './ProjectConfiguration';
import { apiVersions } from './sdkVersion';

import getFileFromStream from './testUtils/getFileFromStream';
import getJSONFileFromStream from './testUtils/getJSONFileFromStream';
import getVinylContents from './testUtils/getVinylContents';
import makeReadStream from './testUtils/makeReadStream';

jest.mock('./packageVersion.const');

const buildId = '0x0f75775f470c1585';
const makeClockfaceProjectConfig = (): ClockProjectConfiguration => ({
  appUUID: 'b4ae822e-eca9-4fcb-8747-217f2a1f53a1',
  appType: AppType.CLOCKFACE,
  appDisplayName: 'My App',
  i18n: {
    'en-US': { name: 'My App' },
    'fr-FR': { name: 'Mon application' },
  },
  buildTargets: ['higgs'],
  requestedPermissions: [],
  defaultLanguage: 'en-US',
});

const makeAppProjectConfig = (): AppProjectConfiguration => ({
  ...makeClockfaceProjectConfig(),
  appType: AppType.APP,
  wipeColor: '#ffaabb',
  iconFile: 'resources/icon.png',
});

let buildStream: Readable;

function addBuildFile(path: string, content: string | Buffer, props: object) {
  buildStream.push(
    new Vinyl({
      path,
      contents: Buffer.isBuffer(content) ? content : Buffer.from(content),
      ...props,
    }),
  );
}

function expectManifestJSON(stream: Duplex) {
  return expect(getJSONFileFromStream(stream, 'manifest.json'));
}

function makeDeviceManifestStream(
  projectConfig: ProjectConfiguration = makeClockfaceProjectConfig(),
) {
  buildStream.push(null);
  return buildStream.pipe(
    makeDeviceManifest({
      buildId,
      projectConfig,
      targetDevice: 'mira',
    }),
  );
}

function makeCompanionManifestStream(
  hasSettings = false,
  projectConfig: ProjectConfiguration = makeClockfaceProjectConfig(),
) {
  buildStream.push(null);
  return buildStream.pipe(
    makeCompanionManifest({
      buildId,
      projectConfig,
      hasSettings,
    }),
  );
}

beforeEach(() => {
  advanceTo(new Date(Date.UTC(2018, 5, 27, 0, 0, 0)));
  buildStream = makeReadStream();
});

it('emits an error if no device entry point is present', () =>
  expectManifestJSON(makeDeviceManifestStream()).rejects.toMatchSnapshot());

it('emits an error if multiple device entry points are present', () => {
  addBuildFile('device/index_A.js', 'foo', {
    isEntryPoint: true,
    componentType: ComponentType.DEVICE,
  });
  addBuildFile('device/index_B.js', 'foo', {
    isEntryPoint: true,
    componentType: ComponentType.DEVICE,
  });
  return expectManifestJSON(
    makeDeviceManifestStream(),
  ).rejects.toMatchSnapshot();
});

it('emits an error if an unrecognised entry point is present when expecting a device entry point', () => {
  addBuildFile('toaster/index.js', 'foo', {
    isEntryPoint: true,
    componentType: 'toaster',
  });
  return expectManifestJSON(
    makeDeviceManifestStream(),
  ).rejects.toMatchSnapshot();
});

it('emits an error if a companion entry point is present when expecting a device entry point', () => {
  addBuildFile('toaster/index.js', 'foo', {
    isEntryPoint: true,
    componentType: ComponentType.COMPANION,
  });
  return expectManifestJSON(
    makeDeviceManifestStream(),
  ).rejects.toMatchSnapshot();
});

describe('when there is a device entry point present', () => {
  beforeEach(() => {
    addBuildFile('device/index.js', 'foo', {
      isEntryPoint: true,
      componentType: ComponentType.DEVICE,
    });
  });

  it('builds a device manifest for a clock', () =>
    expectManifestJSON(makeDeviceManifestStream()).resolves.toMatchSnapshot());

  it('includes supported screen sizes for a clock', () => {
    const projectConfig = {
      ...makeClockfaceProjectConfig(),
      enableProposedAPI: true,
    } as ProjectConfiguration;
    expectManifestJSON(
      makeDeviceManifestStream(projectConfig),
    ).resolves.toMatchSnapshot();
  });

  it('builds a device manifest for an app', () =>
    expectManifestJSON(
      makeDeviceManifestStream(makeAppProjectConfig()),
    ).resolves.toMatchSnapshot());

  it('sets apiVersion in app manifest', () =>
    expectManifestJSON(makeDeviceManifestStream()).resolves.toHaveProperty(
      'apiVersion',
      apiVersions({}).deviceApi,
    ));

  describe('when there are compiled language files', () => {
    beforeEach(() => {
      addBuildFile('lang/english', 'foo', { translationLanguage: 'en-US' });
      addBuildFile('spanish/language', 'foo', { translationLanguage: 'es-ES' });
    });

    it('sets the i18n[lang].resources key for language files that pass through', () =>
      expectManifestJSON(
        makeDeviceManifestStream(),
      ).resolves.toMatchSnapshot());

    it.each(['es-ES', 'en-US'])(
      'ensures the default language %s is the first key in the i18n object',
      (defaultLanguage) => {
        buildStream.push(null);
        const manifest = makeDeviceManifestStream({
          ...makeClockfaceProjectConfig(),
          defaultLanguage,
        });

        return expect(
          getFileFromStream(manifest, 'manifest.json').then(getVinylContents),
        ).resolves.toMatchSnapshot();
      },
    );

    it('passes all files through', (done) => {
      const files: string[] = [];

      makeDeviceManifestStream()
        .on('error', done.fail)
        .on('data', (file: Vinyl) => files.push(file.relative))
        .on('end', () => {
          expect(files).toEqual([
            path.normalize('device/index.js'),
            path.normalize('lang/english'),
            path.normalize('spanish/language'),
            path.normalize('manifest.json'),
          ]);
          done();
        });
    });
  });
});

it('emits an error if no companion entry point is present', () =>
  expectManifestJSON(makeCompanionManifestStream()).rejects.toMatchSnapshot());

it('emits an error if multiple companion entry points are present', () => {
  addBuildFile('companion/index_A.js', 'foo', {
    isEntryPoint: true,
    componentType: ComponentType.COMPANION,
  });
  addBuildFile('companion/index_B.js', 'foo', {
    isEntryPoint: true,
    componentType: ComponentType.COMPANION,
  });
  return expectManifestJSON(
    makeCompanionManifestStream(),
  ).rejects.toMatchSnapshot();
});

it('emits an error if multiple settings entry points are present', () => {
  addBuildFile('companion/index.js', 'foo', {
    isEntryPoint: true,
    componentType: ComponentType.COMPANION,
  });
  addBuildFile('settings/index_A.js', 'foo', {
    isEntryPoint: true,
    componentType: ComponentType.SETTINGS,
  });
  addBuildFile('settings/index_B.js', 'foo', {
    isEntryPoint: true,
    componentType: ComponentType.SETTINGS,
  });
  return expectManifestJSON(
    makeCompanionManifestStream(),
  ).rejects.toMatchSnapshot();
});

it('emits an error if an unrecognised entry point is present when expecting a companion/settings entry point', () => {
  addBuildFile('toaster/index.js', 'foo', {
    isEntryPoint: true,
    componentType: 'toaster',
  });
  return expectManifestJSON(
    makeCompanionManifestStream(),
  ).rejects.toMatchSnapshot();
});

it('emits an error if a device entry point is present when expecting a companion/settings entry point', () => {
  addBuildFile('toaster/index.js', 'foo', {
    isEntryPoint: true,
    componentType: ComponentType.DEVICE,
  });
  return expectManifestJSON(
    makeCompanionManifestStream(),
  ).rejects.toMatchSnapshot();
});

describe('when there is a companion entry point present', () => {
  beforeEach(() => {
    addBuildFile('companion/index.js', 'foo', {
      isEntryPoint: true,
      componentType: ComponentType.COMPANION,
    });
  });

  it('builds a companion manifest', () =>
    expectManifestJSON(
      makeCompanionManifestStream(),
    ).resolves.toMatchSnapshot());

  it('sets apiVersion in companion manifest', () =>
    expectManifestJSON(makeCompanionManifestStream()).resolves.toHaveProperty(
      'apiVersion',
      apiVersions({}).companionApi,
    ));

  it('sets app cluster storage related fields if configured', () =>
    expectManifestJSON(
      makeCompanionManifestStream(false, {
        ...makeClockfaceProjectConfig(),
        developerID: 'f00df00d-f00d-f00d-f00d-f00df00df00d',
        appClusterID: 'a.storage.group',
      }),
    ).resolves.toMatchSnapshot());

  it('emits an error if project has settings but no settings entry point', () =>
    expectManifestJSON(
      makeCompanionManifestStream(true),
    ).rejects.toMatchSnapshot());

  describe('when there is a settings entry point present', () => {
    beforeEach(() => {
      addBuildFile('settings/index.js', 'foo', {
        isEntryPoint: true,
        componentType: ComponentType.SETTINGS,
      });
    });

    it('builds a companion manifest with settings', () =>
      expectManifestJSON(
        makeCompanionManifestStream(true),
      ).resolves.toMatchSnapshot());
  });
});
