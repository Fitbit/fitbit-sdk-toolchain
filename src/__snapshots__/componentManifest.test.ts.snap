// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builds a companion manifest 1`] = `
Object {
  "apiVersion": "1.0.0",
  "buildId": "0x0f75775f470c1585",
  "bundleDate": "2018-06-27T00:00:00.000Z",
  "companion": Object {
    "main": "companion.js",
  },
  "manifestVersion": 2,
  "name": "My App",
  "requestedPermissions": Array [],
  "uuid": "b4ae822e-eca9-4fcb-8747-217f2a1f53a1",
}
`;

exports[`builds a companion manifest with settings 1`] = `
Object {
  "apiVersion": "1.0.0",
  "buildId": "0x0f75775f470c1585",
  "bundleDate": "2018-06-27T00:00:00.000Z",
  "companion": Object {
    "main": "companion.js",
  },
  "manifestVersion": 2,
  "name": "My App",
  "requestedPermissions": Array [],
  "settings": Object {
    "main": "settings.js",
  },
  "uuid": "b4ae822e-eca9-4fcb-8747-217f2a1f53a1",
}
`;

exports[`builds a device manifest for a clock 1`] = `
Object {
  "apiVersion": "1.0.0",
  "appManifestVersion": 1,
  "appType": "clockface",
  "buildId": "0x0f75775f470c1585",
  "bundleDate": "2018-06-27T00:00:00.000Z",
  "i18n": Object {
    "en": Object {
      "name": "My App",
    },
    "fr": Object {
      "name": "Mon application",
    },
  },
  "main": "app/index.js",
  "name": "My App",
  "requestedPermissions": Array [],
  "svgMain": "resources/index.gui",
  "svgWidgets": "resources/widgets.gui",
  "uuid": "b4ae822e-eca9-4fcb-8747-217f2a1f53a1",
}
`;

exports[`builds a device manifest for an app 1`] = `
Object {
  "apiVersion": "1.0.0",
  "appManifestVersion": 1,
  "appType": "app",
  "buildId": "0x0f75775f470c1585",
  "bundleDate": "2018-06-27T00:00:00.000Z",
  "i18n": Object {
    "en": Object {
      "name": "My App",
    },
    "fr": Object {
      "name": "Mon application",
    },
  },
  "iconFile": "resources/icon.png",
  "main": "app/index.js",
  "name": "My App",
  "requestedPermissions": Array [],
  "svgMain": "resources/index.gui",
  "svgWidgets": "resources/widgets.gui",
  "uuid": "b4ae822e-eca9-4fcb-8747-217f2a1f53a1",
  "wipeColor": "#ffaabb",
}
`;

exports[`when there are compiled language files ensures the default locale en is the first key in the i18n object 1`] = `"{\\"appManifestVersion\\":1,\\"main\\":\\"app/index.js\\",\\"svgMain\\":\\"resources/index.gui\\",\\"svgWidgets\\":\\"resources/widgets.gui\\",\\"appType\\":\\"clockface\\",\\"i18n\\":{\\"en\\":{\\"name\\":\\"My App\\",\\"resources\\":\\"lang/english\\"},\\"fr\\":{\\"name\\":\\"Mon application\\"},\\"es\\":{\\"resources\\":\\"spanish/language\\"}},\\"apiVersion\\":\\"1.0.0\\",\\"buildId\\":\\"0x0f75775f470c1585\\",\\"bundleDate\\":\\"2018-06-27T00:00:00.000Z\\",\\"uuid\\":\\"b4ae822e-eca9-4fcb-8747-217f2a1f53a1\\",\\"name\\":\\"My App\\",\\"requestedPermissions\\":[]}"`;

exports[`when there are compiled language files ensures the default locale es is the first key in the i18n object 1`] = `"{\\"appManifestVersion\\":1,\\"main\\":\\"app/index.js\\",\\"svgMain\\":\\"resources/index.gui\\",\\"svgWidgets\\":\\"resources/widgets.gui\\",\\"appType\\":\\"clockface\\",\\"i18n\\":{\\"es\\":{\\"resources\\":\\"spanish/language\\"},\\"en\\":{\\"name\\":\\"My App\\",\\"resources\\":\\"lang/english\\"},\\"fr\\":{\\"name\\":\\"Mon application\\"}},\\"apiVersion\\":\\"1.0.0\\",\\"buildId\\":\\"0x0f75775f470c1585\\",\\"bundleDate\\":\\"2018-06-27T00:00:00.000Z\\",\\"uuid\\":\\"b4ae822e-eca9-4fcb-8747-217f2a1f53a1\\",\\"name\\":\\"My App\\",\\"requestedPermissions\\":[]}"`;

exports[`when there are compiled language files sets the i18n[lang].resources key for language files that pass through 1`] = `
Object {
  "apiVersion": "1.0.0",
  "appManifestVersion": 1,
  "appType": "clockface",
  "buildId": "0x0f75775f470c1585",
  "bundleDate": "2018-06-27T00:00:00.000Z",
  "i18n": Object {
    "en": Object {
      "name": "My App",
      "resources": "lang/english",
    },
    "es": Object {
      "resources": "spanish/language",
    },
    "fr": Object {
      "name": "Mon application",
    },
  },
  "main": "app/index.js",
  "name": "My App",
  "requestedPermissions": Array [],
  "svgMain": "resources/index.gui",
  "svgWidgets": "resources/widgets.gui",
  "uuid": "b4ae822e-eca9-4fcb-8747-217f2a1f53a1",
}
`;
